### PR TITLE
Navigation: Add help menu to top search bar

### LINKED ISSUE #53948: Navigation: Add help menu to top navbar

### PR DESCRIPTION
**Pull request by lpskdl**
**What this PR does / why we need it**:
Added help menu to the top search bar. Also added new behaviour to menu item to support custom display or label

Screenshot 2022-09-13 at 09 57 22

**Which issue(s) this PR fixes**:

Fixes #53948  

**Special notes for your reviewer**:
- The implementation is quite the same as the profile menu. The tricky part is how to display the external link icon on the right side part of the `MenuItem`. I did multiple tries trying not to touch the MenuItem implementation but I didn't really like it. So I decided to support `renderCustomLabel` inside `MenuItem` to place the Icon on the right side without breaking any existing behaviour of the `MenuItem`. Let me know if this is fine, happy to change it.

### CODE REVIEW

File: packages/grafana-ui/src/components/Menu/MenuItem.tsx
Comment (torkelo) [inline packages/grafana-ui/src/components/Menu/MenuItem.tsx:45]: think in other places we use something like 
```ts
label: 'string | React.ReactElement | (() => React.ReactElement);

and then use:  ReactUtils.renderOrCallToRender(label);
```

File: packages/grafana-ui/src/components/Menu/MenuItem.tsx
Comment (torkelo) [inline packages/grafana-ui/src/components/Menu/MenuItem.tsx:151]: I would try using label to support this customization use case (just extend the label type)

File: public/app/core/components/AppChrome/TopBar/TopNavBarMenu.tsx
Comment (torkelo) [inline public/app/core/components/AppChrome/TopBar/TopNavBarMenu.tsx:46]: This feels unnessary, can't we use the icon before the text instead? (which the menu component already supports)

File: packages/grafana-ui/src/components/Menu/MenuItem.tsx
Comment (ashharrison90) [reply packages/grafana-ui/src/components/Menu/MenuItem.tsx:151]: yeah +1, i'd just switch label to be `React.ReactNode`. we'll need to do this anyway for translations since lingui passes `<Trans>` elements around instead of strings

File: public/app/core/components/AppChrome/TopBar/TopNavBarMenu.tsx
Comment (torkelo) [inline public/app/core/components/AppChrome/TopBar/TopNavBarMenu.tsx:39]: This is not great (CTRL + click does not work) and you don't see the URL when you hover over the link. Could not find where these help items are defined, would be great if we could make the ones that are links into normal links

File: public/app/core/components/AppChrome/TopBar/TopNavBarMenu.tsx
Comment (lpskdl) [reply public/app/core/components/AppChrome/TopBar/TopNavBarMenu.tsx:46]: Right, this simplify things. However, i used to see the external link icon placed before the text.

Screenshot 2022-09-13 at 13 57 49

File: packages/grafana-ui/src/components/Menu/MenuItem.tsx
Comment (lpskdl) [reply packages/grafana-ui/src/components/Menu/MenuItem.tsx:151]: right. This could be a breaking change if we extend the support of label. I've browsed a little bit in our repo and `label` is being used as a `key` 🙈 . That is the reason I've use a new prop for this

**Comment by ashharrison90** [comment]
@lpskdl just pushed a commit here that i think addresses everything: https://github.com/grafana/grafana/commit/da7d8617dd7ed664cea167385c04f936fc0c9455

didn't push it to this branch as i didn't wanna clutter things too much for you.

the only thing i'm not sure about is removing the `preventDefault` in the `onClick` of `MenuItem`. it was added to fix a bug with `NodeGraph`, so might need to check what bug it was solving/if it's still needed

**Comment by ashharrison90** [comment]
@aocenas would want to check you're happy with the change we've made to `MenuItem`/`useContextMenu` in particular. we've moved the `preventDefault` out of `MenuItem` and called it specifically in the `onClick` passed by `useContextMenu` instead. i _think_ it's functionally equivalent, and i've tested it with the `gdev-testdata` -> `NodeGraph` -> `response` scenario in `Explore`, but would be good to get an extra set of 👀

**Review by ashharrison90 — APPROVED** [review]
looks great to me, but think we should wait for another review before merging 😅

**Comment by lpskdl** [comment]
Hello @torkelo @aocenas , I would like to request if you could give the changes another look. thanks a lot :)

**Comment by torkelo** [comment]
@lpskdl don't merge yet have some tweaks :)

**Comment by torkelo** [comment]
@lpskdl @ashharrison90 

Pushed some tweaks 
* use Menu header prop instead of custom style on MenuItem for header 
* Move subTitle (version) to Help menu header instead of footer 
* Add icon for keyboard link so all items have icon 
* Tweak styles to leverage theme.spacing support multiple parameters (you can call it like theme.spacing(1,2) to generate `8px 16px` string 

Screenshot 2022-09-16 at 11 28 28

**Review by torkelo — APPROVED** [review]
Needs a enterprise change fix (or make id optional again on FooterLink) 


Screenshot 2022-09-16 at 11 33 25

**Comment by ashharrison90** [comment]
@torkelo have readded the stopPropagation on the header (so you can highlight the version and copy/paste without the menu closing)

otherwise looks great 👍 

@lpskdl do you wanna make the corresponding changes in enterprise or shall i? 🤔

File: packages/grafana-ui/src/components/Menu/MenuItem.tsx
Comment (torkelo) [inline packages/grafana-ui/src/components/Menu/MenuItem.tsx:132]: this change looks unnecessary now

**Comment by aocenas** [comment]
@ashharrison90
> @aocenas would want to check you're happy with the change we've made to MenuItem/useContextMenu in particular. we've moved the preventDefault out of MenuItem and called it specifically in the onClick passed by useContextMenu instead. i think it's functionally equivalent, and i've tested it with the gdev-testdata -> NodeGraph -> response scenario in Explore, but would be good to get an extra set of eyes

Sorry for a late reply. So reason this code was there that in general in grafana we have an issue that ctrl-click does not work consistently which btw I am still seeing in the main navigation submenus, that I cannot open then in a separate tab with ctrl-click which is super annoying tbh. This code made this clicks work and they were low in the stack so that you don't have to think about it when using the context menu. So this may work but will work only for node graph now and any other uses of context menu may now loose the ctrl-click functionality.

**Comment by ashharrison90** [comment]
no worries, thanks for checking @aocenas 

when you say main navigation submenus, do you mean the ones off the sidebar here? 
image

if so, we're aware of that bug (https://github.com/grafana/grafana/issues/47514) and it will be fixed as part of the topnav work.

i don't think those changes are needed for ctrl/cmd-click to work properly. for example in top nav now with these changes we're using the `Menu` component directly and ctrl/cmd-click works:


https://user-images.githubusercontent.com/20999846/190622863-48354bd3-ba89-4376-925f-6040c6c06bb7.mp4

**Comment by torkelo** [comment]
I agree with @ashharrison90 here that having the prevent default handling in MenuItemm onClick seem very strange, if it's to support CTRL+click open new tab then just set url and it will be a normal anchor (a) tag and it will work. 

@aocenas if there are special uses of MenuItem that can't use URLs but onClick and still want CTRL+click handling then I think that is the responsibility of the consumer of MenuItem.

**Comment by aocenas** [comment]
@torkelo @ashharrison90 so I am not sure right now what the underlying issue was. We definitely have both URL and onClick handler in Explore which is because click opens a split pane while new tab should just open the single new query. In this case we had to do the special casing for ctrl-click handling and I thought it makes sense in general as I think URL is required anyway? (not 100% sure either). So yeah if you have onClick + URL which sort of do different thing then I assume that is the main problem cause the onClick will overide the URL change.

**Comment by lpskdl** [comment]
> 

Thanks a lot  @torkelo @ashharrison90 I will fix the remaining issue in enterprise.

**Comment by ericmustin** [comment]
@ashharrison90 Apologies for hopping in mid-review (your work largely lgtm!), but re:

> if so, we're aware of that bug (https://github.com/grafana/grafana/issues/47514) and it will be fixed as part of the topnav work. 🥳

Any rough guesses you're able to provide about where this work lands on y'alls roadmap? This bug has been a huge pain for my organization, and we're debating trying non-upstreamable things on  a fork as a workaround, but if y'all are planning to fix in the short term then would much prefer to just wait for that.

No worries if there's no timeline but figured i'd ask. Thanks again for all the great work y'all are doing!

**Comment by ashharrison90** [comment]
@aocenas got it thanks, i can now replicate that behaviour. just pushed a commit that makes sure it's all doing the exact same stuff so Explore will work exactly as before.

you're right in that cases where there's both an `onClick` and a `url` this will be a change in behaviour slightly (in that they'll now both be called), but i think for consistency's sake that's the correct thing to do.

**Comment by ashharrison90** [comment]
@ericmustin we'll be delivering a large navigation update that we're working on right now, it's one of the top priorities for our squad (@grafana/user-essentials). i wouldn't want to commit to any dates on when we'll deliver it as i'm sure you can imagine, but **soon**™️ 

thanks for your interest 🙌

**Pull request merged**
